### ISSUE TITLE
[expo-gl][Android][standalone] gl.texImage2D doesn't render image in Android standalone app if asset localUri path protocol is different from "file://"

### ISSUE THREAD

**Issue by serhiipalash** [expo/expo#2693]
### Environment

| software         | version
| ---------------- | -------
| expo | 34.0.0
| react-native     | https://github.com/expo/react-native/archive/sdk-34.0.0.tar.gz
| device    | Pixel 2 XL (simulator), Samsung Galaxy A8 (real device)
| platform | Android

`exp diagnostics` output

<img width="1052" alt="Screen Shot 2019-08-22 at 2 13 53 AM" src="https://user-images.githubusercontent.com/14208343/63474634-8e145300-c482-11e9-92c5-4303e804c321.png">

### Steps to Reproduce

The bug is only reproducible in Android standalone app. When you test it in development with Expo mobile app all local uris always start with `file://`. 
I used [GLHeadlessRenderingScreen.tsx](https://github.com/expo/expo/blob/a47a147235a4b64230b86da45b8763472c68a18d/apps/native-component-list/src/screens/GL/GLHeadlessRenderingScreen.tsx) module from Expo examples as `App.js` in my test repo. To test the bug do
```
git clone https://github.com/serhiipalash/test-glview-expo
cd test-glview-expo
yarn
expo build:android
```
Download .apk and install it on Android Simulator.

Screen recordings:

Android simulator Pixel 2 XL (in my real device Galaxy A8 behaviour is the same): 
https://drive.google.com/file/d/1OlWgCrb0jZrZYIaO-Wubd3wbkkOKuIN-/view

App doesn't work in this line

<img width="1075" alt="screen shot 2018-11-12 at 2 54 10 pm" src="https://user-images.githubusercontent.com/14208343/48349049-c2551a80-e68b-11e8-97e7-3c8e5b92cf84.png">

because this code in Android standalone app returns local uri which starts with `asset://`

<img width="655" alt="screen shot 2018-11-12 at 3 01 39 pm" src="https://user-images.githubusercontent.com/14208343/48349180-2f68b000-e68c-11e8-8b05-4840862d22d4.png">

<img width="756" alt="screen shot 2018-10-30 at 11 36 47 pm" src="https://user-images.githubusercontent.com/14208343/48349192-368fbe00-e68c-11e8-80d4-c525a5b0b7ce.png">

There is no way to resolve local or remote asset in Android standalone app to local path with `file://` protocol using `Expo.Asset` or `Expo.FileSystem`. 

### Expected Behavior

In Android standalone app you can resolve asset from module and pass it to `gl.texImage2D(...)` and it should be rendered independent from uri path protocol.

### Actual Behavior

In Android standalone app asset resolves to local uri with protocol `asset://` and `gl.texImage2D(...)` doesn't render image for such uri.

I assume that to fix this issue you have to add all missing protocols in `expo-gl` native code here
https://github.com/expo/expo/blob/master/packages/expo-gl-cpp/cpp/UEXGL.cpp

These protocols I found as missing

`data://`
`asset://` - All local paths in Android Expo standalone app
`assets-library://` -CameraRoll.getPhotos iOS
`content://` - CameraRoll.getPhotos Android

### Reproducible Demo

https://github.com/serhiipalash/test-glview-expo

ping @tsapeta

**Comment by serhiipalash** [comment]
Hi everyone!
@EvanBacon, you had an idea why is this bug happens. Maybe need to share it with @tsapeta ?
It would be great to see this bug fix in SDK 32 :)

**Comment by serhiipalash** [comment]
The bug is fixed.
Not sure who fixed it, but thank you @tsapeta !
Our app works properly, but the test repo I created has another bug. The app starts without crashing, but the image is not rendered.

<img width="589" alt="screen shot 2018-12-19 at 7 34 09 pm" src="https://user-images.githubusercontent.com/14208343/50237566-ebbd4080-03c5-11e9-8724-7e1f3fc8f34c.png">

In development mode it looks like this

<img width="633" alt="screen shot 2018-12-19 at 7 33 46 pm" src="https://user-images.githubusercontent.com/14208343/50237580-f7106c00-03c5-11e9-9cca-3392f63c50f2.png">

I guess it is another bug in the code of "Headless GL" example from Expo docs. This code I used for test repo. So be aware that "Headless GL" wont work in Android standalone app.

I am closing this issue. Thanks again!

ping @EvanBacon

**Comment by serhiipalash** [comment]
I've reopened this issue because the bug with not rendering image with path protocol different from `file://` or `https://` is still there. Please check my previous comment to see screenshots.

In Android standalone app when you try to use local bandled asset as source for `gl.texImage2D` it doesn't work because source path is `asset://`. In our app we temporary fixed this by using remote assets because they have protocol `https://`, but it is not a production way as it slows down view rendering compared to using local assets and our app can no longer work offline.

The issue is reproducible in Expo v31. I will update the test repo to Expo v32 in next few days, test and write about result here.

@tsapeta can you please try to fix it?

**Comment by beaurushton** [comment]
I'm also seeing this issue.  Has any fix been found?

**Comment by beaurushton** [comment]
@tsapeta it seems like you might be the wizard who can fix this?

**Comment by beaurushton** [comment]
I've been looking at this for a few days and haven't found any decent solutions.  What I have worked out though, is that when the assets are bundled for offline use (included in the "assetBundlePatterns" in app.json) this is what causes the change to asset:// based localUri paths on android release builds.  

So my current workaround is to not include these assets in the "assetBundlePatterns" but also means, as @serhiipalash mentioned above, is that these assets cannot be used in an offline situation and require a network to load off the expo CDN.

I've looked through RNs asset loading code at https://github.com/facebook/react-native/blob/master/Libraries/Image/AssetSourceResolver.js but haven't been able to come up with any solutions.

@EvanBacon This may also relate to some of the issues logged with the expo-three package such as https://github.com/expo/expo-three/issues/54

**Comment by adbl** [comment]
@serhiipalash @beaurushton I think we are facing this issue too, but we are on Expo 32, are you seeing this issue for that version too? The app has to work offline, so loading from url is not a viable solution in our case. I'm new to Expo and it's troubling to see that such an issue remains after half a year of being reported...

**Comment by adbl** [comment]
@serhiipalash I see you have https://github.com/gre/webgltexture-loader/pull/8 also. Seems related, does it fix this somehow?

**Comment by serhiipalash** [comment]
> I see you have [gre/webgltexture-loader#8](https://github.com/gre/webgltexture-loader/pull/8) also. Seems related, does it fix this somehow?

@adbl no it doesn't fix this issue. Pull request to `webgltexture-loader` fixes `GLImage` React-component logic that resolves if passed uri is local or remote. But this issue is about `gl.texImage2D` method and not about `GLImage` component. So, there were two issues that prevent rendering local images using `GLImage` component, first in `GLImage` and second in `gl.texImage2D`. The first one I fixed (you can use my fork of `gl-react-expo`) , but the second issues is still there.

**Comment by serhiipalash** [comment]
> I think we are facing this issue too, but we are on Expo 32, are you seeing this issue for that version too?

Yes we are on version 32 and the issue is there too.

**Comment by adbl** [comment]
Ok, thanks. What I'm currently attempting is to do something like 
```
const asset = Asset.fromModule(module);
await asset.downloadAsync()
// now use `asset.localUri` as source
```

Even though these assets are already bundled, we'll see. Have you tried something like that?

Docs say:
_
- localUri
If the asset has been downloaded (by calling downloadAsync()), the file:// URI pointing to the local file on the device that contains the asset data.
_

**Comment by serhiipalash** [comment]
> Docs say:
> 
> * localUri
>   If the asset has been downloaded (by calling downloadAsync()), the file:// URI pointing to the local file on the device that contains the asset data.

Docs are wrong. `asset.downloadAsync()` will return local uri with `asset://` protocol in Android standalone expo-app and this local uri can't be supplied to `gl.texImage2D`.

What you can try is to replace `asset://` on `file://` in local uri string. In Android Expo-client app this will return uri with `file://` protocol
```
const asset = Asset.fromModule(module);
await asset.downloadAsync()
```
and everything will work correctly.

**Comment by adbl** [comment]
Didn't work, because bundled assets will have a `localUri` of `asset:///asset_fb2dc6e0e10bdb7d83314ee669ed0ed9.png`, contradictory to the docs 😕

**Comment by adbl** [comment]
Are you saying to try changing the `asset://...` above to be `file://...` instead? I’m probably misunderstanding this, could you elaborate the workaround a bit? @serhiipalash

**Comment by beaurushton** [comment]
@adbl I tried replacing `asset://...` with `file://...` in the string too but couldn't get that to work either.  I still don't have a solution other than excluding the assets from my asset bundle pattern and downloading and caching them from the CDN.

**Comment by adbl** [comment]
Here is our current workaround for this issue:

```

// bundled asset uri's become `asset://...` in production build, but expo-gl
// cannot handle them, https://github.com/expo/expo/issues/2693
// this workaround copies them to a known path files so we can use a regular
// `file://...` uri instead.
async function copyAssetToCacheAsync(assetModule, localFilename) {
  const localUri = `${FileSystem.cacheDirectory}asset_${localFilename}`;
  const fileInfo = await FileSystem.getInfoAsync(localUri, { size: false });
  if (!fileInfo.exists) {
    const asset = Asset.fromModule(assetModule);
    await asset.downloadAsync();
    console.log(`copyAssetToCacheAsync ${asset.localUri} -> ${localUri}`);
    await FileSystem.copyAsync({
      from: asset.localUri,
      to: localUri,
    });
  }
  return localUri;
}
```

All the assets needed will first go through this thing before I can render anything. then I'll use the return value like `const funnyImg = copyAssetToCacheAsync(require('...'), 'funny')`. Lame, but works.

**Comment by danmaas** [comment]
I hit this issue today and worked around it using the hack here:
https://github.com/expo/expo-three/issues/104#issuecomment-498989294

I spect the root causes is here, in the native C++ implementation of expo-gl:
https://github.com/expo/expo/blob/master/packages/expo-gl-cpp/cpp/UEXGL.cpp

The `loadImage()` function has a special case to handle Assets passed in the form of JavaScript objects with a .localUri property. However, it only handles the file:// prefix and fails to handle the asset:// prefix (which only appears for assets loaded in standalone Android builds).

**Comment by serhiipalash** [comment]
> in the native C++ implementation of expo-gl:

@danmaas can you create a pull request to fix its logic? You can check my pull request to `webgltexture-loader-expo` and see all cases you need to add https://github.com/gre/webgltexture-loader/pull/8/files#diff-b17c0f21d84380de2e11010f680d2230

I am not familiar with C++ and can't do it by myself.

**Comment by serhiipalash** [comment]
> module.uri.startsWith("/") // Expo.takeSnapshotAsync in DEV in Expo 31

This one should be already fixed in the latest Expo (always starts with `file:`), but I think you can assign @tsapeta on your pull request as a reviewer and you will know this for sure.

**Comment by serhiipalash** [comment]
I've updated test repo for this issue to Expo v34 and the bug is still there (v34.0.4), but now `gl.textImage2D` just doesn't render the image without crashing the app.

https://github.com/serhiipalash/test-glview-expo

I've also updated this issue description and recorded a new video how this bug looks today.

https://drive.google.com/file/d/1OlWgCrb0jZrZYIaO-Wubd3wbkkOKuIN-/view

**Comment by serhiipalash** [comment]
And here the Android build I made, if someone from Expo team wants to test it too https://expo.io/builds/bd87c3f0-3f91-4e4c-a94e-455a3948097f

**Comment by danmaas** [comment]
I encountered yet another flavor of this issue today when I switched my project from managed to bare workflow.

In bare Android builds, drawable images resolve to asset.localUris that look like `"assets_images_myimagefile"`. Note that there is no "asset:" prefix and the file format suffix (`.png` in my case) has been stripped.

`https://github.com/expo/expo/blob/master/packages/expo-gl-cpp/cpp/UEXGL.cpp` loadImage() fails to produce a valid OpenGL texture because it has no special-case handling for this new form of asset.localUri reference.

I worked around this problem by adding `https://github.com/itinance/react-native-fs` to my project and using its `readFileRes()` function to copy the bytes of the drawable image into a local file (using expo-file-system `writeAsStringAsync()`), and then passing the "file://" URI into expo-gl's texture loader.

Basically, there's an architectural issue here where `expo-asset` has added a bunch of different "flavors" of asset reference beyond just "file://", but the `expo-gl` stack has not been updated to handle any of these new forms of asset storage.

**Comment by mtsap** [comment]
> I encountered yet another flavor of this issue today when I switched my project from managed to bare workflow.
> 
> In managed, standalone Android builds, my texture image asset.localUris look like `"asset:///asset_images_myimagefile"` and are handled by the copyAssetToCacheAsync() work-around I posed above.
> 
> Now, in bare Android builds, the same images resolve to asset.localUris that look like `"assets_images_myimagefile"`. Note that there is no "asset:" prefix and the file format suffix (`.png` in my case) has been stripped.
> 
> Consequently, in
> https://github.com/expo/expo/blob/master/packages/expo-gl-cpp/cpp/UEXGL.cpp
> loadImage() fails to load the image into an OpenGL texture because it has no special-case handling for this new form of asset.localUri reference.
> 
> I worked around this problem by adding https://github.com/itinance/react-native-fs to my bare builds and using its `readFileRes()` function to copy the bytes of the drawable image into a local file (using expo-file-system `writeAsStringAsync()`), and then passing the "file://" URI into expo-gl's texture loader.
> 
> Basically, the root problem seems to be an architectural issue where `expo-asset` has now added a bunch of different "flavors" of asset reference beyond just `"file://"`, but the `expo-gl` stack has not been updated to handle any of these new ways to reference image files.
> 
> TypeScript is not alerting about this problem because the asset.localUri references are all just strings, which masks the fact that different formats need different, platform-specific handling in the texture loading code.

Had the exact same issue. Used RNFS to copy the image to the Documents directory. Appended the `file://` extension to it and put it in an object with localUri property. Passed it to gl.textImage2D and it worked.

**Comment by pacoelayudante** [comment]
I found out that the specific file that filters the uri is 
https://github.com/expo/expo/blob/3d9ef386bde5911d15306a46b50e7b1bcea3abf5/packages/expo-gl-cpp/cpp/EXGLContext.cpp#L82

Maybe that can help?

**Comment by wood1986** [comment]
why it has not been fixed for a while?

**Comment by pacoelayudante** [comment]
I did a weird fix like this for my case, don't know if might interest you
https://forums.expo.io/t/asset-to-file/35206/5

**Comment by serhiipalash** [comment]
I've just tested our app with latest version of all packages and the issue is still there. The dependencies are

```
    "expo-file-system": "~9.3.0",
    "expo-gl": "^10.0.0",
    "expo-gl-cpp": "^10.0.0",
    "gl-react": "^5.0.0",
    "gl-react-native": "^5.0.0",
    "react": "16.14.0",
    "react-native": "0.63.4",
    "react-native-unimodules": "^0.12.0",
```

This fix https://github.com/expo/expo/issues/2693#issuecomment-488631357 still helps to fix the issue

**Comment by f4z3k4s** [comment]
Sadly, [above solution](https://github.com/expo/expo/issues/2693#issuecomment-488631357) is still required today because of the [explanation here](https://github.com/expo/expo/issues/2693#issuecomment-539218518) for plain png files. I want to pass the uri to a native function and call `new URL(uri);` on it, but urls that are like `"assets_images_myimagefile"` and have no `file://` are invalid.
